## `Task` Do feature selection as per metods taught is session 54 on SECOM dataset.

Dataset Link : https://archive.ics.uci.edu/ml/datasets/SECOM

Drive Link : https://docs.google.com/spreadsheets/d/1dFCe1zgokabsiEr6BbWmMJtiMefkrChpJWLiG_0dDkk/edit?usp=share_link

In [69]:
# Write your Code here

### `Solution`

In [70]:
import pandas as pd
import numpy as np

data = pd.read_csv("https://docs.google.com/spreadsheets/d/e/2PACX-1vQtBXo5cBnDsM2fmfHPm6u72KGUS5FjPHNGMxOfYjA9-CAhmnRpwkIw_rOR3sANJIToiUU__6fbBvig/pub?gid=572763137&single=true&output=csv")

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1567 entries, 0 to 1566
Columns: 592 entries, Time to Pass/Fail
dtypes: float64(590), int64(1), object(1)
memory usage: 7.1+ MB


In [71]:
data.head()

Time        0        1  ...     588       589  Pass/Fail
0  2008-07-19 11:55:00  3030.93  2564.00  ...     NaN       NaN         -1
1  2008-07-19 12:32:00  3095.78  2465.14  ...  0.0060  208.2045         -1
2  2008-07-19 13:17:00  2932.61  2559.94  ...  0.0148   82.8602          1
3  2008-07-19 14:43:00  2988.72  2479.90  ...  0.0044   73.8432         -1
4  2008-07-19 15:22:00  3032.24  2502.87  ...  0.0044   73.8432         -1

[5 rows x 592 columns]

In [72]:
# Logistic Model Before Feature Selection
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


# Dropping Time Column
data = data.drop('Time', axis=1)

# Filling NaN Values with Random
for column in data.columns:
    # Print("Column-", column)
    # Get the minimum and maximum values of the column
    min_value = data[column].min()
    max_value = data[column].max()
    
    # Generate Random numbers within the range
    random_values = np.random.uniform(min_value, max_value, size=data[column].isnull().sum())
    
    # Create a Series with the Random Values
    random_series = pd.Series(random_values, index=data[column][data[column].isnull()].index)
    
    # Fill NaN values with the random Series
    data[column].fillna(random_series, inplace=True)
    
    # Print
    # print(data[column].isnull().sum())
    
# data.isnull().sum()

C:\Users\ACER\AppData\Local\Temp\ipykernel_4780\3241618202.py:24: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data[column].fillna(random_series, inplace=True)


In [73]:
# Separate Features and Target
X = data.drop('Pass/Fail', axis=1)
y = data['Pass/Fail']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(X_train.shape)
print(X_test.shape)


# Initialize and Train Logistic regression Model
log_reg = LogisticRegression(max_iter=10000) # Increase max_iter if it does not converge
log_reg.fit(X_train, y_train)

# Make prediction on the test set
y_pred = log_reg.predict(X_test)

# Calculate and Print Accuracy Score
accuracy = accuracy_score(y_test, y_pred)
print("Test Accuracy:", accuracy)

(1253, 590)
(314, 590)
Test Accuracy: 0.9044585987261147


C:\Users\ACER\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


To perform filter-based feature selection on the "UCI SECOM" dataset, which has 592 columns and a target column called "Pass/Fail," we can utilize the following methods:

1. Duplicate Features:
   - Identify and remove duplicate columns from the dataset. Columns with identical values provide redundant information and do not contribute to the prediction task.

2. Variance Threshold Method:
   - Calculate the variance of each feature.
   - Remove features with low variance, as they tend to have little or no predictive power.
   - Set a threshold value for variance and remove features below that threshold.

3. Correlation:
   - Compute the correlation matrix of the features.
   - Identify highly correlated features and choose one from each highly correlated group.
   - High correlation between features indicates redundancy, and removing one from each correlated group helps reduce multicollinearity.

4. ANOVA (Analysis of Variance):
   - Perform an ANOVA test between each feature and the target variable ("Pass/Fail").
   - Select features with a significant impact on the target variable.
   - Set a significance level (e.g., p-value threshold) for the test to determine the importance of each feature.

5. Chi-Squared:
   - Apply the Chi-Squared test between each feature and the target variable, considering both features as categorical.
   - Select features with a significant association with the target variable.
   - Set a significance level (e.g., p-value threshold) to determine the importance of each feature.

Implementing these feature selection methods in Python using the "UCI SECOM" dataset can be done as follows:

In [74]:
import pandas as pd
import numpy as np
from sklearn.feature_selection import VarianceThreshold, SelectKBest, chi2, f_classif
from scipy.stats import pearsonr

data_path = "https://docs.google.com/spreadsheets/d/e/2PACX-1vQtBXo5cBnDsM2fmfHPm6u72KGUS5FjPHNGMxOfYjA9-CAhmnRpwkIw_rOR3sANJIToiUU__6fbBvig/pub?gid=572763137&single=true&output=csv"

# Load the DataSet
DATA = pd.read_csv(data_path)

# Remove Duplicate Features
# Get the Subset of columns with Duplicate Values
duplicated_cols = DATA.columns[DATA.T.duplicated()]

# Remove the Duplicated Columns
data = DATA.drop(columns=duplicated_cols)

# Drop Time Column
data.drop('Time', inplace=True, axis=1)

# Number of Columns after removing Duplicate Columns
print("Number of Columns - ", DATA.shape[1])
print("Number of Columns after removing duplicate columns- ", data.shape[1])

Number of Columns -  592
Number of Columns after removing duplicate columns-  487


In [75]:
# Variance Threshold Method
selector = VarianceThreshold(threshold=0.01)
sel = selector.fit(data)

columns = data.columns[sel.get_support()]

data_vt = sel.transform(data)

data_vt = pd.DataFrame(data_vt, columns=columns)

# Number of Columns after Variance Threshold Method
print("Number of Columns after Variance Threshold Method- ", data_vt.shape[1])

Number of Columns after Variance Threshold Method-  315


In [76]:
# Correlation

corr_matrix = data_vt.corr().abs()
upper_tri = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool_))
to_drop = [column for column in upper_tri.columns if any(upper_tri[column] > 0.8)]

data_corr = data_vt.drop(to_drop, axis=1)

# Number of Columns after Variance Threshold Method
print("Number of Columns after Correlation- ", data_corr.shape[1])

Number of Columns after Correlation-  162


In [77]:
data_corr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1567 entries, 0 to 1566
Columns: 162 entries, 0 to Pass/Fail
dtypes: float64(162)
memory usage: 1.9 MB


ANOVA approach can not be done here because there are NaN values inside this dataset. But we can still use ANOVA to each column individually, comparing the target variable which is Pass/Fail column in our case against the non-missing values in that specific column.

We will drop NaN values of each column before performing the ANOVA

In [78]:
from scipy.stats import f_oneway

# Significance Value
alpha = 0.05

# Columns having p_value less than Alpha
column_pvalues = []

# Iterate over each column
for column in data_corr.iloc[:, :-1].columns:
    # Extract the non-missing values in the column
    column_data = data_corr[column].dropna()
    
    # Perform ANOVA with the Target Variable
    anova_result = f_oneway(column_data, data_corr['Pass/Fail'])
    
    # Print the ANOVA results or perform further analysis
    print(f"Column: {column} - ANOVA p-value: {anova_result.pvalue}")
    
    if anova_result.pvalue <= alpha:
        column_pvalues.append((column, anova_result.pvalue))

# Selecting best 100 Features - Lower P-Value means better feature
# Sort the Column P-Value in Ascending Order
column_pvalues.sort(key=lambda x: x[1])

# Select the Top 100 Columns with the Lowest P-Values
selected_columns = [column for column, _ in column_pvalues[:100]]

data_anova = data_corr[selected_columns+['Pass/Fail']]

print("Number of Columns after Correlation- ", data_anova.shape[1])

Column: 0 - ANOVA p-value: 0.0
Column: 1 - ANOVA p-value: 0.0
Column: 2 - ANOVA p-value: 0.0
Column: 3 - ANOVA p-value: 0.0
Column: 4 - ANOVA p-value: 0.0003805229753154665
Column: 6 - ANOVA p-value: 0.0
Column: 12 - ANOVA p-value: 0.0
Column: 14 - ANOVA p-value: 0.0
Column: 15 - ANOVA p-value: 0.0
Column: 16 - ANOVA p-value: 0.0
Column: 18 - ANOVA p-value: 0.0
Column: 19 - ANOVA p-value: 0.0
Column: 21 - ANOVA p-value: 0.0
Column: 22 - ANOVA p-value: 0.0
Column: 23 - ANOVA p-value: 0.0
Column: 24 - ANOVA p-value: 5.0211194207923185e-05
Column: 25 - ANOVA p-value: 0.0
Column: 28 - ANOVA p-value: 0.0
Column: 29 - ANOVA p-value: 0.0
Column: 31 - ANOVA p-value: 0.0
Column: 32 - ANOVA p-value: 0.0
Column: 33 - ANOVA p-value: 0.0
Column: 34 - ANOVA p-value: 0.0
Column: 35 - ANOVA p-value: 0.0
Column: 37 - ANOVA p-value: 0.0
Column: 38 - ANOVA p-value: 0.0
Column: 39 - ANOVA p-value: 0.0
Column: 40 - ANOVA p-value: 0.0
Column: 41 - ANOVA p-value: 0.0
Column: 43 - ANOVA p-value: 0.0
Column: 4

Chi-Squared cant be done because our Data is Numerical.

In [79]:
# Shape of Data after Feature Selection
print("Shape - ", data_anova.shape)

# Filling NaN values with Random
for column in data_anova.columns:
    
    # Get the minimum and maximum values of the colum
    min_value = data_anova[column].min()
    max_value = data_anova[column].max()
    
    # Generate Random numbers within the range
    random_values = np.random.uniform(min_value, max_value, size=data_anova[column].isnull().sum())
    
    # Create a Series with the Random Values
    random_series = pd.Series(random_values, index=data_anova[column][data_anova[column].isnull()].index)
    
    # Fill NaN values with the random Series
    data_anova[column].fillna(random_series, inplace=True)
    
    # Print
    # print(data_anova[column].isnull().sum())

# Seperate features and target
X = data_anova.drop("Pass/Fail", axis=1)
y = data_anova["Pass/Fail"]
    
# Split Data into Training and Testing data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
print(X_train.shape)
print(X_test.shape)
    
    
# Initialize and Train Logistic regression Model
log_reg = LogisticRegression(max_iter=10000) # Increase max_iter if it does not converge
log_reg.fit(X_train, y_train)

# Make prediction on the test set
y_pred = log_reg.predict(X_test)

# Calculate and Print Accuracy Score
accuracy = accuracy_score(y_test, y_pred)
print("Test Accuracy:", accuracy)

Shape -  (1567, 101)


C:\Users\ACER\AppData\Local\Temp\ipykernel_4780\3246670434.py:18: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data_anova[column].fillna(random_series, inplace=True)
C:\Users\ACER\AppData\Local\Temp\ipykernel_4780\3246670434.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_anova[column].fillna(random_series, inplace=True)


(1253, 100)
(314, 100)
Test Accuracy: 0.9203821656050956


C:\Users\ACER\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
